File: python3/sdk/ais_sdk/asr_sentence.py
# -*- coding:utf-8 -*-
import ssl
import signer
from urllib.error import URLError, HTTPError
import urllib.parse
import urllib.request
import json

#
# access asr, asr_sentence,post data by token
#
def asr_sentence(token, data, url, encode_type='wav', sample_rate='8k'):
    _url = 'https://ais.cn-north-1.myhuaweicloud.com/v1.0/voice/asr/sentence'

    if data != '':
        data = data.decode("utf-8")

    _data = {
      "url":url,
      "data": data,
      "encode_type": encode_type,
      "sample_rate": sample_rate
    }

    _headers = {
        "Content-Type": "application/json",
        "X-Auth-Token": token
    }
    data = json.dumps(_data).encode("utf-8")
    kreq = urllib.request.Request(_url, data, _headers)
    resp = None
    status_code = None
    try:
        # 
        # Here we use the unvertified-ssl-context, Because in FunctionStage
        # the client CA-validation have some problem, so we must do this.
        #
        _context = ssl._create_unverified_context()
        r = urllib.request.urlopen(kreq,context=_context)
        
    #
    # We use HTTPError and URLError，because urllib can't process the 4XX &
    # 500 error in the single urlopen function.
    #
    # If you use a modern, high-level designed HTTP client lib, Yeah, I mean requests, 
    # there is no this problem. 
    #
    except HTTPError as e:
        resp = e.read()
        status_code = e.code
    except URLError as e:
        resp = e.read()
        status_code = e.code
    else:
        status_code = r.code
        resp = r.read()
    return resp.decode('utf-8')

#
# access asr, asr_sentence,post data by ak,sk
#
def asr_sentence_aksk(_ak, _sk, data, url, encode_type='wav', sample_rate='8k'):
    _url = "https://ais.cn-north-1.myhuaweicloud.com/v1.0/voice/asr/sentence"

    sig = signer.Signer()
    sig.AppKey = _ak
    sig.AppSecret = _sk

    if data != '':
        data = data.decode("utf-8")

    _data = {
        "url": url,
        "data": data,
        "encode_type": encode_type,
        "sample_rate": sample_rate
    }

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = "ais.cn-north-1.myhuaweicloud.com"
    kreq.uri = "/v1.0/voice/asr/sentence"
    kreq.method = "POST"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.body = json.dumps(_data)

    resp = None
    status_code = None
    try:
        sig.Sign(kreq)
        #
        # Here we use the unvertified-ssl-context, Because in FunctionStage
        # the client CA-validation have some problem, so we must do this.
        #
        _context = ssl._create_unverified_context()
        req = urllib.request.Request(url=_url, data=bytes(kreq.body), headers=kreq.headers)
        r = urllib.request.urlopen(req,context=_context)
    #
    # We use HTTPError and URLError，because urllib2 can't process the 4XX &
    # 500 error in the single urlopen function.
    #
    # If you use a modern, high-level designed HTTP client lib, Yeah, I mean requests,
    # there is no this problem.
    #
    except HTTPError as e:
        resp = e.read()
        status_code = e.code
    except URLError as e:
        resp = e.read()
        status_code = e.code
    else:
        status_code = r.code
        resp = r.read()
    return resp.decode('utf-8')

File: php/ais_sdk/asr_sentence.php
<?php
require "signer.php";
require "ais.php";

/**
 * token 方式
 */
function asr_sentence($token, $data, $url = "", $encode_type = "wav", $sample_rate = "16k")
{

    // 构建请求信息
    $_url = "https://" . ASR_ENDPOINT . ASR_SENTENCE;

    $data = array(
        "data" => $data,                        // data: 音频文件的base64
        "url" => $url,                          // url: 音频文件的obs的url地址
        "encode_type" => $encode_type,          // encode_type：支持语音的格式
        "sample_rate" => $sample_rate,          // sample_rate：语音的采样率，目前支持16k，代表16HZ
    );

    $curl = curl_init();
    $headers = array(
        "Content-Type:application/json",
        "X-Auth-Token:" . $token);

    // 请求信息封装
    curl_setopt($curl, CURLOPT_URL, $_url);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($data));
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, TRUE);
    curl_setopt($curl, CURLOPT_NOBODY, FALSE);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);

    // 执行请求信息
    $response = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    if ($status == 0) {
        echo curl_error($curl);
    } else {

        // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
        if ($status == 200) {
            return json_encode(json_decode($response, true),JSON_UNESCAPED_UNICODE);
        } else {
            echo "Http status is: " . $status . "\n";
            echo $response;
        }

    }
    curl_close($curl);


}

/**
 * ak,sk 方式
 */
function asr_sentence_aksk($_ak, $_sk, $data, $url = "", $encode_type = "wav", $sample_rate = "16k")
{
    // 构建ak，sk对象
    $signer = new Signer();
    $signer->AppKey = $_ak;             // 构建ak
    $signer->AppSecret = $_sk;          // 构建sk

    // 构建请求对象
    $req = new Request();
    $req->method = "POST";
    $req->scheme = "https";
    $req->host = ASR_ENDPOINT;
    $req->uri = ASR_SENTENCE;

    $data = array(
        "data" => $data,                        // data: 音频文件的base64
        "url" => $url,                          // url: 音频文件的obs的url地址
        "encode_type" => $encode_type,          // encode_type：支持语音的格式
        "sample_rate" => $sample_rate,          // sample_rate：语音的采样率，目前支持16k，代表16HZ
    );

    $headers = array(
        "Content-Type" => "application/json",
    );

    $req->headers = $headers;
    $req->body = json_encode($data);

    // 获取ak，sk方式的请求对象，执行请求
    $curl = $signer->Sign($req);
    curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
    $response = curl_exec($curl);
    $status = curl_getinfo($curl, CURLINFO_HTTP_CODE);
    if ($status == 0) {
        echo curl_error($curl);
    } else {

        // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
        if ($status == 200) {
            return json_encode(json_decode($response, true),JSON_UNESCAPED_UNICODE);
        } else {

            echo "Http status is: " . $status . "\n";
            echo $response;
        }

    }
    curl_close($curl);
}
